exchange/exchange.go

```diff
@@ -139,20 +139,31 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
 	conversions := e.currencyConverter.Rates()
 
 	adapterBids, adapterExtra := e.getAllBids(auctionCtx, cleanRequests, aliases, bidAdjustmentFactors, blabels, conversions)
-	bidCategory, adapterBids, err := applyCategoryMapping(ctx, requestExt, adapterBids, *categoriesFetcher, targData)
-	auc := newAuction(adapterBids, len(bidRequest.Imp))
-	if err != nil {
-		return nil, fmt.Errorf("Error in category mapping : %s", err.Error())
+	anyBidsReturned := false
+	for _, seatBid := range adapterBids {
+		if seatBid != nil && len(seatBid.bids) > 0 {
+			anyBidsReturned = true
+			break
+		}
 	}
```

Comment:
Maybe we can save a bit of processing time if we remove the `for` statement in line 143 and make the `getAllBids()` method to return the `anyBidsReturned` value. Since the `adapterBids` map gets already traversed inside `getAllBids()`,  we could save some processing time by not going through the rather large `adapterBids` map again and make our solution a bit less verbose. 
In other words, go from this:
```
141 adapterBids, adapterExtra := e.getAllBids(auctionCtx, cleanRequests, aliases, bidAdjustmentFactors, blabels, conversions)
142 anyBidsReturned := false
143 for _, seatBid := range adapterBids {
144 	if seatBid != nil && len(seatBid.bids) > 0 {
145 		anyBidsReturned = true
146 		break
147 	}
148 }
149 if anyBidsReturned {
```
To something like this:
```
141 adapterBids, adapterExtra, anyBidsReturned := e.getAllBids(auctionCtx, cleanRequests, aliases, bidAdjustmentFactors, blabels, conversions)
142 if anyBidsReturned {
```

Inside of `getAllBids()` we can probably populate the boolean value in the last if statement:
Go from:
```
183 func (e *exchange) getAllBids(ctx context.Context, cleanRequests map[openrtb_ext.BidderName]*openrtb.BidRequest, aliases map[string]string, bidAdjustments map[string]float64, blabels map[openrtb_ext.BidderName]*pbsmetrics.AdapterLabels, conversions currencies.Conversions) (map[openrtb_ext.BidderName]*pbsOrtbSeatBid, map[openrtb_ext.BidderName]*seatResponseExtra) {
184     // Set up pointers to the bid results
185     adapterBids := make(map[openrtb_ext.BidderName]*pbsOrtbSeatBid, len(cleanRequests))
186     adapterExtra := make(map[openrtb_ext.BidderName]*seatResponseExtra, len(cleanRequests))
187     chBids := make(chan *bidResponseWrapper, len(cleanRequests))
188
189     for bidderName, req := range cleanRequests {
190         // Here we actually call the adapters and collect the bids.
191 +-- 45 lines: coreBidder := resolveBidder(string(bidderName), aliases)--------------------------------------------
236     }
237     // Wait for the bidders to do their thing
238     for i := 0; i < len(cleanRequests); i++ {
239         brw := <-chBids
240         adapterBids[brw.bidder] = brw.adapterBids
241         adapterExtra[brw.bidder] = brw.adapterExtra
242     }
243
244     return adapterBids, adapterExtra
245 }
```
To something like:
```
176 func (e *exchange) getAllBids(ctx context.Context, cleanRequests map[openrtb_ext.BidderName]*openrtb.BidRequest, aliases map[string]string, bidAdjustments map[string]float64, blabels map[openrtb_ext.BidderName]*pbsmetrics.AdapterLabels, conversions currencies.Conversions) (map[openrtb_ext.BidderName]*pbsOrtbSeatBid, map[openrtb_ext.BidderName]*seatResponseExtra, bool) {
177     // Set up pointers to the bid results
178     adapterBids := make(map[openrtb_ext.BidderName]*pbsOrtbSeatBid, len(cleanRequests))
179     adapterExtra := make(map[openrtb_ext.BidderName]*seatResponseExtra, len(cleanRequests))
180     bidsFound := false  // NEW
181     chBids := make(chan *bidResponseWrapper, len(cleanRequests))
182
183     for bidderName, req := range cleanRequests {
184         // Here we actually call the adapters and collect the bids.
185 +-- 45 lines: coreBidder := resolveBidder(string(bidderName), aliases)---------------------------------
230     }
231     // Wait for the bidders to do their thing
232     for range cleanRequests {
233         //for i := 0; i < len(cleanRequests); i++ {
234         brw := <-chBids
235         adapterBids[brw.bidder] = brw.adapterBids
236         if adapterBids[brw.bidder] != nil && len(adapterBids[brw.bidder].bids) > 0 { // NEW
237             bidsFound = true        // NEW
238         }                           // NEW
239         adapterExtra[brw.bidder] = brw.adapterExtra
240     }
241
242     return adapterBids, adapterExtra, bidsFound
243 }
```
What do you think? Please let me know if you agree with any of this.

Reply:
I think it's a good idea! Refactored.